### PR TITLE
Only load manifest once within the dataset and share Manifest amount the readers

### PR DESCRIPTION
**Pull request by eddyxu**
Additionally it only reads the dictionary columns once , to prevent repeated read dictionary on each open file.

### CODE REVIEW

**Comment by eddyxu** [comment]
After this PR, seems the FileReader::Open in `s3sdk` is the most expensive operation

Before the PR
Screenshot 2022-09-08 at 4 54 21 PM

After this PR
Screenshot 2022-09-09 at 3 42 33 PM

File: cpp/src/lance/io/reader_test.cc [hunk 1]
```diff
@@ -12,7 +12,7 @@
 //  See the License for the specific language governing permissions and
 //  limitations under the License.
 
-#include "lance/arrow/reader.h"
+#include "lance/io/reader.h"
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader_test.cc:15]: why move this?

File: cpp/src/lance/format/schema.h [hunk 2]
```diff
@@ -232,6 +232,7 @@ class Field final {
   friend class FieldVisitor;
   friend class ToArrowVisitor;
   friend class WriteDictionaryVisitor;
+  friend class LoadDictionaryVisitor;
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.h:235]: What's the difference between Read and Load?

File: cpp/src/lance/format/schema.cc [hunk 3]
```diff
@@ -345,6 +346,7 @@ std::shared_ptr<Field> Field::Copy(bool include_children) const {
   new_field->logical_type_ = logical_type_;
   new_field->extension_name_ = extension_name_;
   new_field->encoding_ = encoding_;
+  new_field->dictionary_ = dictionary_;
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:349]: this is for dictionary arrays? or is this some other dictionary?

File: cpp/src/lance/io/reader_test.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader_test.cc:15]: I removed the public FileReader interface. We did not use it publicly before, and it only acts a proxy to `lance::io::FileReader`. 

So this PR needs to pass Schema / Manifest to FileReader, if we are still using `lance::arrow::FileReader` it means that we need to public `Schema / Manifest` as well. I feel that it is not necessary in the moment.

File: cpp/src/lance/format/schema.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.h:235]: Oh, this is loading the string values of the dictionary. I am fine to rename it.  wdyt.

File: cpp/src/lance/format/schema.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:349]: Yes, this is the dictionary value array (i.e., strings), do not contain indices.

File: cpp/src/lance/format/schema.h (hunk 2 above)
Comment (changhiskhan) [reply cpp/src/lance/format/schema.h:235]: Let's just be consistent if it's doing the same thing

**Review by changhiskhan — APPROVED** [review]
one small nit in the comments

File: cpp/src/lance/format/schema.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.h]: fixed.

**Pull request merged**